File: java/components/validation-ctt/HEAD/ctt_testcases_v15.04.00a/src/com/at4wireless/alljoyn/core/commons/ServiceAvailabilityHandler.java
/*******************************************************************************
 *  Copyright AllSeen Alliance. All rights reserved.
 *
 *     Permission to use, copy, modify, and/or distribute this software for any
 *     purpose with or without fee is hereby granted, provided that the above
 *     copyright notice and this permission notice appear in all copies.
 *
 *     THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 *     WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 *     MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 *     ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 *     WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 *     ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 *     OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *******************************************************************************/
package com.at4wireless.alljoyn.core.commons;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

import org.alljoyn.services.common.ServiceAvailabilityListener;

// TODO: Auto-generated Javadoc
/**
 * The Class ServiceAvailabilityHandler.
 */
public class ServiceAvailabilityHandler implements ServiceAvailabilityListener
{
    
    /** The count down latch. */
    private  CountDownLatch countDownLatch = new CountDownLatch(1);

    /* (non-Javadoc)
     * @see org.alljoyn.services.common.ServiceAvailabilityListener#connectionLost()
     */
    @Override
    public void connectionLost()
    {
        countDownLatch.countDown();
    }

    /**
     * Wait for session lost.
     *
     * @param timeout the timeout
     * @param unit the unit
     * @return true, if successful
     * @throws InterruptedException the interrupted exception
     */
    public  boolean waitForSessionLost(long timeout, TimeUnit unit) throws InterruptedException
    {
        return countDownLatch.await(timeout, unit);
    }

}

File: cpp/components/validation-ctt/HEAD/ctt_testcases/ServiceAvailabilityHandler.cpp
/******************************************************************************
* Copyright (c) 2016 Open Connectivity Foundation (OCF) and AllJoyn Open
*    Source Project (AJOSP) Contributors and others.
*
*    SPDX-License-Identifier: Apache-2.0
*
*    All rights reserved. This program and the accompanying materials are
*    made available under the terms of the Apache License, Version 2.0
*    which accompanies this distribution, and is available at
*    http://www.apache.org/licenses/LICENSE-2.0
*
*    Copyright 2016 Open Connectivity Foundation and Contributors to
*    AllSeen Alliance. All rights reserved.
*
*    Permission to use, copy, modify, and/or distribute this software for
*    any purpose with or without fee is hereby granted, provided that the
*    above copyright notice and this permission notice appear in all
*    copies.
*
*     THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL
*     WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED
*     WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE
*     AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL
*     DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR
*     PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER
*     TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
*     PERFORMANCE OF THIS SOFTWARE.
******************************************************************************/
#include "stdafx.h"
#include "ServiceAvailabilityHandler.h"

#include <thread>

void ServiceAvailabilityHandler::SessionLost(ajn::SessionId t_SessionId, SessionLostReason t_Reason)
{
	LOG(INFO) << "SessionLost sessionId = " << t_SessionId << ", Reason = " << t_Reason;
	--m_countdown_latch;
}

bool ServiceAvailabilityHandler::waitForSessionLost(const long t_Timeout)
{
	const clock_t beginTime = clock();

	while (m_countdown_latch == 1 && (clock() - beginTime < t_Timeout))
	{
		std::this_thread::sleep_for(std::chrono::milliseconds(100));
	}

	return m_countdown_latch == 0;
}